kotest-extensions/kotest-extensions-junit5extensions/src/jvmMain/kotlin/io/kotest/extensions/junit5/KotestExtensionContext.kt

```diff
@@ -0,0 +1,86 @@
+package io.kotest.extensions.junit5
+
+import io.kotest.core.spec.Spec
+import io.kotest.core.spec.description
+import org.junit.jupiter.api.TestInstance
+import org.junit.jupiter.api.extension.ExtensionContext
+import org.junit.jupiter.api.extension.TestInstances
+import java.lang.reflect.AnnotatedElement
+import java.lang.reflect.Method
+import java.util.Optional
+import java.util.function.Function
+
+class KotestExtensionContext(private val spec: Spec) : ExtensionContext {
+
+   override fun getParent(): Optional<ExtensionContext> = Optional.empty()
+   override fun getRoot(): ExtensionContext = this
+
+   override fun getUniqueId(): String = spec::class.description().id()
+
+   override fun getDisplayName(): String = spec::class.description().fullName()
+
+   override fun getTags(): MutableSet<String> = spec.tags().map { it.name }.toMutableSet()
+
+   override fun getElement(): Optional<AnnotatedElement> = Optional.empty()
+
+   override fun getTestClass(): Optional<Class<*>> = Optional.of(spec::class.java)
+   override fun getTestMethod(): Optional<Method> = Optional.empty()
+
+   override fun getTestInstanceLifecycle(): Optional<TestInstance.Lifecycle> =
+      Optional.of(TestInstance.Lifecycle.PER_CLASS)
+
+   override fun getTestInstance(): Optional<Any> = Optional.of(spec)
+   override fun getTestInstances(): Optional<TestInstances> = Optional.of(KotestTestInstances(spec))
+
+   override fun getExecutionException(): Optional<Throwable> = Optional.empty()
+   override fun getConfigurationParameter(key: String?): Optional<String> = Optional.empty()
+
+   override fun publishReportEntry(map: MutableMap<String, String>?) {}
+
+   override fun getStore(namespace: ExtensionContext.Namespace?): ExtensionContext.Store {
+      return ExtensionStore()
+   }
+}
+
+class KotestTestInstances(private val instance: Spec) : TestInstances {
+   override fun getInnermostInstance(): Any = instance
+   override fun getEnclosingInstances(): MutableList<Any> = mutableListOf(instance)
+   override fun getAllInstances(): MutableList<Any> = mutableListOf(instance)
+   override fun <T : Any?> findInstance(requiredType: Class<T>?): Optional<T> = Optional.of(instance as T)
+}
+
+class ExtensionStore : ExtensionContext.Store {
+
+   private val map = mutableMapOf<Any?, Any?>()
+
+   override fun get(key: Any?): Any? = map[key]
+
+   override fun <V : Any?> get(key: Any?, requiredType: Class<V>?): V? {
+      return map[key] as? V
+   }
```

Comment:
Type casting will here. I have tried out with following test case.

```kotlin
class ExtensionStoreTest : FunSpec({
   data class Person(val name: String, val age: Int)
   data class Vehicle(val name: String, val model: String)

   test("should give null when value for given is not of given type in store") {
      val extensionStore = ExtensionStore()
      val person = Person("foo", 20)

      extensionStore.put("foo", person)

      val vehicle = extensionStore.get("foo", Vehicle::class.java)

      vehicle shouldBe null
   }
})
```

I think here expectation was to get null, but we get type case error.

```bash
java.lang.ClassCastException: class io.kotest.extensions.junit5.ExtensionStoreTest$1$Person cannot be cast to class io.kotest.extensions.junit5.ExtensionStoreTest$1$Vehicle (io.kotest.extensions.junit5.ExtensionStoreTest$1$Person and io.kotest.extensions.junit5.ExtensionStoreTest$1$Vehicle are in unnamed module of loader 'app')
	at io.kotest.extensions.junit5.ExtensionStoreTest$1$1.invokeSuspend(ExtensionStoreTest.kt:16)
	at io.kotest.extensions.junit5.ExtensionStoreTest$1$1.invoke(ExtensionStoreTest.kt)
	at io.kotest.core.runtime.ExecutionsKt$executeWithBehaviours$2$1.invokeSuspend(executions.kt:16)
	at io.kotest.core.runtime.ExecutionsKt$executeWithBehaviours$2$1.invoke(executions.kt)
	at io.kotest.core.runtime.ExecutionsKt.wrapTestWithGlobalAssert(executions.kt:43)
	at io.kotest.core.runtime.ExecutionsKt$executeWithBehaviours$2.invokeSuspend(executions.kt:15)
	at io.kotest.core.runtime.ExecutionsKt$executeWithBehaviours$2.invoke(executions.kt)
```

Reply:
I checked the junit default implementation and they throw when the type is not right.

I've added your test, changed it to expect a throw and tidied up the code, including adding in namespace support.